frontend/src/services/HTTPHelper.ts

```diff
@@ -0,0 +1,368 @@
+    this.options = buildOptions(options);
+  }
+
+  public tryGet<TResult>(
+    url: string | URL,
+    request: HTTPGETRequest,
+    cancellation?: CancellationToken
+  ): Promise<HTTPResponse<TResult>> {
+    return this.executeRequest(url, request, 'get', undefined, cancellation);
+  }
+
+  public tryPost<TResult>(
+    url: string | URL,
+    request: HTTPPOSTRequest,
+    cancellation?: CancellationToken
+  ): Promise<HTTPResponse<TResult>> {
+    return this.executeRequest(
+      url,
+      request,
+      'post',
+      request.body,
+      cancellation
+    );
+  }
+
+  private executeRequest<TResult>(
+    url: string | URL,
+    request: HTTPRequest,
+    method: 'post' | 'get' | 'head' | 'put' | 'delete' | 'patch',
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any, @typescript-eslint/explicit-module-boundary-types
+    body?: any,
+    cancellation?: CancellationToken
+  ): Promise<HTTPResponse<TResult>> {
+    // Pre-process arguments, extract the header collection and set a default content type header, that can be overridden by the caller.
+    const headers: HTTPHeaderCollection = {
+      'Content-Type': 'application/json',
+      ...request.headers,
+    };
+
+    const parsedURL = this.appendQuery(url, request.query);
+
+    return new Promise<HTTPResponse<TResult>>((resolve, reject) => {
+      // If the cancellation token is already canceled, we can reject right away.
+      if (cancellation?.isCancellationRequested) {
+        reject(new CancellationError());
+      }
+
+      const httpClient = new XMLHttpRequest();
+      httpClient.open(method, parsedURL.href, true);
+
+      // Process the request-headers
+      // Walk over all attributes of the 'headers' object and set the key-value pairs as request header.
+      const keys = Object.keys(headers);
+      for (let i = 0; i < keys.length; i += 1) {
+        const key = keys[i];
+        const value = headers[key];
+
+        if (typeof value === 'string') {
+          httpClient.setRequestHeader(key, value);
+        } else {
+          throw new ArgumentError(
+            'The header collection is not a valid set of key-value pairs.',
+            'headers'
+          );
+        }
+      }
+
+      // Register to the cancellation token, if present.
+      let cancellationUnsubscribe = nop;
+      if (cancellation) {
+        cancellationUnsubscribe = cancellation.subscribe(() =>
+          httpClient.abort()
+        );
+      }
+
+      // Callback that is executed, when the load operation was performed.
+      // This does not necessarily mean, that everything was successful.
+      httpClient.onload = () => {
+        // Unsubscribe from the cancellation token.
+        cancellationUnsubscribe();
+
+        let result: TResult | null = null;
+
+        // Check the HTTP response status code to be in the success range.
+        if (httpClient.status >= 200 && httpClient.status <= 299) {
+          result = JSON.parse(httpClient.response);
```

Comment:
Will fail when header `'Content-Type': 'application/json'` is overwritten.

Reply:
That's true, but what's the alternative? A generic value formatter that can be injection? That would be overkill. If the caller wants to shoot themselves in the foot... go for it